master: add document for partial index (#21903)
pingcap/docs: sql-statements/sql-statement-create-index.md

- If a table uses multi-valued indexes, you cannot back up, replicate, or import the table using BR, TiCDC, or TiDB Lightning to a TiDB cluster earlier than v6.6.0.
- For a query with complex conditions, TiDB might not be able to select multi-valued indexes. For information on the condition patterns supported by multi-valued indexes, refer to [Use multi-valued indexes](/choose-index.md#use-multi-valued-indexes).

## Partial indexes <span class="version-mark">New in v8.5.7 and v9.0.0</span>

> A partial index is an index built on a subset of rows in a table. When creating a partial index, you can specify a conditional expression, also known as a predicate, to define the subset. The index contains entries only for rows that satisfy the predicate.

Reviewer: ```suggestion
A partial index is an index built on a subset of rows in a table. When creating a partial index, you can specify a conditional expression, also known as a predicate, to define that subset of rows. The index contains entries only for the rows that satisfy the predicate.
```

- If a table uses multi-valued indexes, you cannot back up, replicate, or import the table using BR, TiCDC, or TiDB Lightning to a TiDB cluster earlier than v6.6.0.
- For a query with complex conditions, TiDB might not be able to select multi-valued indexes. For information on the condition patterns supported by multi-valued indexes, refer to [Use multi-valued indexes](/choose-index.md#use-multi-valued-indexes).

> ## Partial indexes

Reviewer 2: ```suggestion
## Partial indexes <span class="version-mark">New in v8.5.7</span>
```

> 

Reviewer 2: ```suggestion

Starting from v8.5.7, TiDB supports creating partial indexes.

```

> A partial index is an index built on a subset of rows in a table, defined by a conditional expression (called the predicate of the partial index). The index contains entries only for those rows that satisfy the predicate.

Reviewer 2: ```suggestion
A partial index is an index built on a subset of rows in a table. A conditional expression, called a predicate, defines the subset. The index contains entries only for rows that satisfy the predicate.
```

### Create partial indexes

You can create partial indexes by adding a `WHERE` clause to the index definition. For example:

```sql
CREATE TABLE t1 (c1 INT, c2 INT, c3 TEXT);
CREATE INDEX idx1 ON t1 (c1) WHERE c2 > 10;
```

You can also create partial indexes using `ALTER TABLE`:

```sql
ALTER TABLE t1 ADD INDEX idx2 (c1, c2) WHERE c3 = 'abc';
```

> Or specify the partial index when creating the table:

Reviewer 2: ```suggestion
You can also specify a partial index when you create the table:
```

```sql
CREATE TABLE t2 (
    id INT PRIMARY KEY,
    status VARCHAR(20),
    created_at DATETIME,
    INDEX idx_active_status (status) WHERE status = 'active'
);
```

### Usage examples

The following examples demonstrate how to use partial indexes effectively:

```sql
-- Create a table with user data
CREATE TABLE users (
    id INT PRIMARY KEY AUTO_INCREMENT,
    name VARCHAR(100),

>     status varchar(20),

Reviewer 2: ```suggestion
    status VARCHAR(20),
```

    created_at DATETIME,
    score INT
);

-- Create partial indexes for common query patterns
CREATE INDEX idx_active_users ON users (name) WHERE status = 'active';
CREATE INDEX idx_high_score_users ON users (created_at) WHERE score > 1000;
CREATE INDEX idx_pending_status ON users (created_at) WHERE status = 'pending';
```

Then the following queries can use the partial index:

> ```

Reviewer 2: ```suggestion
```sql
```

> mysql> explain SELECT * FROM users WHERE status = 'active' AND name = 'John';

Reviewer 2: ```suggestion
mysql> EXPLAIN SELECT * FROM users WHERE status = 'active' AND name = 'John';
```

| id                            | estRows | task      | access object                             | operator info                                         |
+-------------------------------+---------+-----------+-------------------------------------------+-------------------------------------------------------+
| IndexLookUp_9                 | 1.00    | root      |                                           |                                                       |
| ├─IndexRangeScan_6(Build)     | 10.00   | cop[tikv] | table:users, index:idx_active_users(name) | range:["John","John"], keep order:false, stats:pseudo |
| └─Selection_8(Probe)          | 1.00    | cop[tikv] |                                           | eq(test.users.status, "active")                       |
|   └─TableRowIDScan_7          | 10.00   | cop[tikv] | table:users                               | keep order:false, stats:pseudo                        |
+-------------------------------+---------+-----------+-------------------------------------------+-------------------------------------------------------+
4 rows in set (0.00 sec)

> mysql> explain SELECT * FROM users WHERE status = 'active' ORDER BY name;

Reviewer 2: ```suggestion
mysql> EXPLAIN SELECT * FROM users WHERE status = 'active' ORDER BY name;
```

+-------------------------------+----------+-----------+-------------------------------------------+---------------------------------+
| id                            | estRows  | task      | access object                             | operator info                   |
+-------------------------------+----------+-----------+-------------------------------------------+---------------------------------+
| IndexLookUp_18                | 10.00    | root      |                                           |                                 |
| ├─IndexFullScan_15(Build)     | 10000.00 | cop[tikv] | table:users, index:idx_active_users(name) | keep order:true, stats:pseudo   |
| └─Selection_17(Probe)         | 10.00    | cop[tikv] |                                           | eq(test.users.status, "active") |
|   └─TableRowIDScan_16         | 10000.00 | cop[tikv] | table:users                               | keep order:false, stats:pseudo  |
+-------------------------------+----------+-----------+-------------------------------------------+---------------------------------+
4 rows in set (0.00 sec)

> mysql> explain SELECT * FROM users WHERE score > 10000 ORDER BY created_at;

Reviewer 2: ```suggestion
mysql> EXPLAIN SELECT * FROM users WHERE score > 10000 ORDER BY created_at;
```

+-------------------------------+----------+-----------+-----------------------------------------------------+--------------------------------+
| id                            | estRows  | task      | access object                                       | operator info                  |
+-------------------------------+----------+-----------+-----------------------------------------------------+--------------------------------+
| IndexLookUp_18                | 3333.33  | root      |                                                     |                                |
| ├─IndexFullScan_15(Build)     | 10000.00 | cop[tikv] | table:users, index:idx_high_score_users(created_at) | keep order:true, stats:pseudo  |
| └─Selection_17(Probe)         | 3333.33  | cop[tikv] |                                                     | gt(test.users.score, 10000)     |
|   └─TableRowIDScan_16         | 10000.00 | cop[tikv] | table:users                                         | keep order:false, stats:pseudo |
+-------------------------------+----------+-----------+-----------------------------------------------------+--------------------------------+
4 rows in set (0.00 sec)

> mysql> explain SELECT * FROM users WHERE status = 'pending';

Reviewer 2: ```suggestion
mysql> EXPLAIN SELECT * FROM users WHERE status = 'pending';
```

+-------------------------+----------+-----------+---------------+----------------------------------+
| id                      | estRows  | task      | access object | operator info                    |
+-------------------------+----------+-----------+---------------+----------------------------------+
| TableReader_8           | 10.00    | root      |               | data:Selection_7                 |
| └─Selection_7           | 10.00    | cop[tikv] |               | eq(test.users.status, "pending") |
|   └─TableFullScan_6     | 10000.00 | cop[tikv] | table:users   | keep order:false, stats:pseudo   |
+-------------------------+----------+-----------+---------------+----------------------------------+
3 rows in set (0.00 sec)
```

> If the predicates in query don't meet the index definition, the index will not be chosen even with hint:

Reviewer 2: ```suggestion
If the query predicates do not match the partial index definition, TiDB does not select the index, even with a hint:
```

```

> mysql> explain SELECT * FROM users use index(idx_high_score_users) WHERE score > 100 ORDER BY created_at;

Reviewer 2: ```suggestion
mysql> EXPLAIN SELECT * FROM users USE INDEX(idx_high_score_users) WHERE score > 100 ORDER BY created_at;
```

+---------------------------+----------+-----------+---------------+--------------------------------+
| id                        | estRows  | task      | access object | operator info                  |
+---------------------------+----------+-----------+---------------+--------------------------------+
| Sort_5                    | 3333.33  | root      |               | test.users.created_at          |
| └─TableReader_11          | 3333.33  | root      |               | data:Selection_10              |
|   └─Selection_10          | 3333.33  | cop[tikv] |               | gt(test.users.score, 100)      |
|     └─TableFullScan_9     | 10000.00 | cop[tikv] | table:users   | keep order:false, stats:pseudo |
+---------------------------+----------+-----------+---------------+--------------------------------+
```

### When to use partial indexes

Partial indexes are particularly useful in the following scenarios:

- **Selective filtering**: When you frequently query a small subset of rows based on specific conditions

> - **Conditional uniqueness**: When you need unique constraints that only apply under certain conditions

Reviewer 2: ```suggestion
- **Selective filtering**: use a partial index when you frequently query a small subset of rows based on specific conditions.
- **Conditional uniqueness**: use a partial index when you need unique constraints that apply only under certain conditions.
```

### Limitations

> - The `WHERE` clause in partial indexes supports basic comparison operators (`=`, `!=`, `<`, `<=`, `>`, `>=`) and `IN` predicates with constant values

Reviewer 3: IS (NOT) NULL is not mentioned?

- The types of the column and constant value should be the same
- The predicate can only reference columns from the same table

> - Partial indexes cannot be used on expression indexes

Reviewer 2: ```suggestion
- The `WHERE` clause in partial indexes supports basic comparison operators (`=`, `!=`, `<`, `<=`, `>`, `>=`) and `IN` predicates with constant values.
- The column and constant value must have the same type.
- The predicate can only reference columns from the same table.
- Partial indexes cannot be used on expression indexes.
```

### Performance benefits

> Partial indexes offer several advantages:

Reviewer 2: ```suggestion
Partial indexes have the following performance benefits:
```

1. **Reduced storage**: Only rows matching the predicate are indexed, saving storage space

> 2. **Faster DML**: It'll be faster to maintain the index of a subset of data during INSERT, UPDATE, and DELETE operations

Reviewer 3: Something like `3. Ability to maintain a unique constraint on a subset of the table rows`?

Author: It has been mentioned in the `When to use partial indexes` section. It's not part of the `Performance benefits`.

> 2. **Faster DML**: It'll be faster to maintain the index of a subset of data during INSERT, UPDATE, and DELETE operations

Reviewer 2: ```suggestion
- **Reduced storage**: only the rows that match the predicate are indexed, which saves storage space.
- **Faster DML**: maintaining an index for a subset of rows is faster during `INSERT`, `UPDATE`, and `DELETE` operations.
```

- If a table uses multi-valued indexes, you cannot back up, replicate, or import the table using BR, TiCDC, or TiDB Lightning to a TiDB cluster earlier than v6.6.0.
- For a query with complex conditions, TiDB might not be able to select multi-valued indexes. For information on the condition patterns supported by multi-valued indexes, refer to [Use multi-valued indexes](/choose-index.md#use-multi-valued-indexes).

## Partial indexes <span class="version-mark">New in v8.5.7</span>

> Starting from v8.5.7, TiDB supports creating partial indexes.

Reviewer: ```suggestion
## Partial indexes <span class="version-mark">New in v8.5.7 and v9.0.0</span>
```

|   └─TableRowIDScan_16         | 10000.00 | cop[tikv] | table:users                                         | keep order:false, stats:pseudo |
+-------------------------------+----------+-----------+-----------------------------------------------------+--------------------------------+
4 rows in set (0.00 sec)

mysql> EXPLAIN SELECT * FROM users WHERE status = 'pending';
+-------------------------+----------+-----------+---------------+----------------------------------+
| id                      | estRows  | task      | access object | operator info                    |
+-------------------------+----------+-----------+---------------+----------------------------------+
| TableReader_8           | 10.00    | root      |               | data:Selection_7                 |
| └─Selection_7           | 10.00    | cop[tikv] |               | eq(test.users.status, "pending") |
|   └─TableFullScan_6     | 10000.00 | cop[tikv] | table:users   | keep order:false, stats:pseudo   |
+-------------------------+----------+-----------+---------------+----------------------------------+
3 rows in set (0.00 sec)
```

> If the query predicates do not match the partial index definition, TiDB does not select the index, even with a hint:

Reviewer: ```suggestion
If a query predicate does not match the partial index definition, TiDB does not select the partial index, even with a hint. For example, the following statement cannot use the partial index `idx_high_score_users`, because the query predicate `score > 100` does not match the partial index definition `score > 1000`:
```

+---------------------------+----------+-----------+---------------+--------------------------------+
| id                        | estRows  | task      | access object | operator info                  |
+---------------------------+----------+-----------+---------------+--------------------------------+
| Sort_5                    | 3333.33  | root      |               | test.users.created_at          |
| └─TableReader_11          | 3333.33  | root      |               | data:Selection_10              |
|   └─Selection_10          | 3333.33  | cop[tikv] |               | gt(test.users.score, 100)      |
|     └─TableFullScan_9     | 10000.00 | cop[tikv] | table:users   | keep order:false, stats:pseudo |
+---------------------------+----------+-----------+---------------+--------------------------------+
```

### When to use partial indexes

Partial indexes are particularly useful in the following scenarios:

- **Selective filtering**: use a partial index when you frequently query a small subset of rows based on specific conditions.
- **Conditional uniqueness**: use a partial index when you need unique constraints that apply only under certain conditions.

### Limitations

> - The `WHERE` clause in partial indexes supports basic comparison operators (`=`, `!=`, `<`, `<=`, `>`, `>=`), `IS NULL`, `IS NOT NULL` and `IN` predicates with constant values.

Reviewer: ```suggestion
- The `WHERE` clause in partial indexes supports basic comparison operators (`=`, `!=`, `<`, `<=`, `>`, `>=`), `IS NULL`, `IS NOT NULL`, and `IN` predicates with constant values.
```

> - The column and constant value must have the same type.

Reviewer: ```suggestion
- The column and constant values must be of the same data type.
```

- The predicate can only reference columns from the same table.

> - Partial indexes cannot be used on expression indexes.

Reviewer: question: used on or created on?

Author: Created on. I'll modify it.
